Fix four bugs with numeric dimension output types.

gianm opened the pull request:
This patch includes the following bug fixes:

- TopNColumnSelectorStrategyFactory: Cast dimension values to the output type
  during dimExtractionScanAndAggregate instead of updateDimExtractionResults.
  This fixes a bug where, for example, grouping on doubles-cast-to-longs would
  fail to merge two doubles that should have been combined into the same long value.
- TopNQueryEngine: Use DimExtractionTopNAlgorithm when treating string columns
  as numeric dimensions. This fixes a similar bug: grouping on string-cast-to-long
  would fail to merge two strings that should have been combined.
- GroupByQuery: Cast numeric types to the expected output type before comparing them
  in compareDimsForLimitPushDown. This fixes #6123.
- GroupByQueryQueryToolChest: Convert Jackson-deserialized dimension values into
  the proper output type. This fixes an inconsistency between results that came
  from cache vs. not-cache: for example, Jackson sometimes deserializes integers
  as Integers and sometimes as Longs.

And the following code-cleanup changes, related to the fixes above:

- DimensionHandlerUtils: Introduce convertObjectToType, compareObjectsAsType,
  and converterFromTypeToType to make it easier to handle casting operations.
- TopN in general: Rename various "dimName" variables to "dimValue" where they
  actually represent dimension values. The old names were confusing.

leventov commented on processing/src/main/java/io/druid/query/topn/PooledTopNAlgorithm.java:
```diff
@@ -749,14 +743,9 @@ protected void updateResults(
-        }
-
-
+        // Output type must be STRING in order for PooledTopNAlgorithm to make sense; so no need to convert value.
```
Maybe add an assert about the output type here.

leventov commented on processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java:
```diff
@@ -19,34 +19,36 @@
 {
-  @Override
-  public int getCardinality(DimensionSelector selector)
+  private final Function<Object, Comparable<?>> dimensionValueConverter;
```
Function from String?

leventov commented on processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java:
```diff
@@ -126,7 +119,7 @@ private long dimExtractionScanAndAggregateWithCardinalityKnown(
       DimensionSelector selector,
       Aggregator[][] rowSelector,
-      Map<String, Aggregator[]> aggregatesStore
+      Map<Comparable, Aggregator[]> aggregatesStore
```
`Comparable` and `Comparable<?>` are used inconsistently in this class

leventov commented on processing/src/main/java/io/druid/query/topn/types/TopNColumnSelectorStrategyFactory.java:
```diff
@@ -27,23 +28,39 @@
-        return new NumericTopNColumnSelectorStrategy.OfDouble();
+        if (ValueType.isNumeric(dimensionType)) {
+          // Return strategy that aggregates using the _output_ type, because this allows us to collapse values
+          // properly (numeric types cannot represent all values of other numeric types).
```
FWIW `double` represents all values of `float`

leventov commented on processing/src/main/java/io/druid/query/topn/types/TopNColumnSelectorStrategyFactory.java:
```diff
@@ -27,23 +28,39 @@
+          // properly (numeric types cannot represent all values of other numeric types).
+          return NumericTopNColumnSelectorStrategy.ofType(dimensionType, dimensionType);
+        } else {
+          // Return strategy that aggregates using the _input_ type. Here we are assuming that the output type can
```
Despite those comments, it's still hard to understand what is going on here. Maybe you could try to clarify it more?

leventov commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -302,6 +306,64 @@ public static Float convertObjectToFloat(@Nullable Object valObj, boolean report
+
+  @Nullable
+  public static Comparable<?> convertObjectToType(
+      @Nullable final Object obj,
```
Unnecessary breakdown

leventov commented:
@fjy please don't merge PRs with unanswered comments (despite the approval), otherwise why they are left?

gianm commented:
Fwiw, I had already started addressing some of those comments, so I plan to raise a follow up PR for that.

Btw, @leventov what did you intend by doing an approval + also leaving comments? I'd interpret that as "please consider these comments, but if you don't want to do them, I am ok with that." I thought they were all reasonable comments so that's why I'm doing a follow up.

gianm commented:
Follow-up is in #6231.
